azure/services/vmextensions/vmextensions.go

```diff
@@ -56,47 +61,45 @@ func (s *Service) Name() string {
 
 // Reconcile creates or updates the VM extension.
 func (s *Service) Reconcile(ctx context.Context) error {
-	ctx, log, done := tele.StartSpanWithLogger(ctx, "vmextensions.Service.Reconcile")
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "vmextensions.Service.Reconcile")
 	defer done()
 
-	for _, extensionSpec := range s.Scope.VMExtensionSpecs() {
-		if existing, err := s.client.Get(ctx, s.Scope.ResourceGroup(), extensionSpec.VMName, extensionSpec.Name); err == nil {
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
+
+	specs := s.Scope.VMExtensionSpecs()
+	if len(specs) == 0 {
+		return nil
+	}
+
+	// We go through the list of ExtensionSpecs to reconcile each one, independently of the result of the previous one.
+	// If multiple errors occur, we return the most pressing one.
+	//  Order of precedence (highest -> lowest) is: error that is not an operationNotDoneError (i.e. error creating) -> operationNotDoneError (i.e. creating in progress) -> no error (i.e. created)
+	var resultErr error
+	for _, extensionSpec := range specs {
+		result, err := s.CreateResource(ctx, extensionSpec, serviceName)
+		if err != nil {
+			if !azure.IsOperationNotDoneError(err) || resultErr == nil {
+				resultErr = err
+			}
+		} else {
+			vmextension, ok := result.(compute.VirtualMachineExtension)
+			if !ok {
+				return errors.Errorf("%T is not a compute.VirtualMachineExtension", result)
+			}
+
 			// check the extension status and set the associated conditions.
-			if retErr := s.Scope.SetBootstrapConditions(ctx, to.String(existing.ProvisioningState), extensionSpec.Name); retErr != nil {
+			if retErr := s.Scope.SetBootstrapConditions(ctx, to.String(vmextension.ProvisioningState), extensionSpec.ResourceName()); retErr != nil {
+				// TODO: what precedence should this error have?
```

Comment:
`SetBootstrapConditions` might need a bit of refactoring. Instead of taking the error and using that to set the condition like we're doing for other resources, it's taking the provisioning state of the extension and using that to set the condition, then returning an error if the extension is done creating. Can we change it to always run (even if resultErr is not nil) and use the error to determine what the condition should be set to instead? Similar to what we're doing in `UpdatePutStatus`